### ISSUE TITLE
Nested catch rethrows the parent exception in async methods

### ISSUE THREAD

**Issue by ThomasGoulet73** [dotnet/roslyn#71569]
**Version Used**: 
Does not reproduce when built with MSBuild 17.8.3+195e7f5a3
Reproduces when built with MSBuild 17.9.0-preview-23578-04+5fcddc790

**Steps to Reproduce**:
Build this repro with a recent version of MSBuild and run it. The Test1 method shows the bug and Test2 method shows that the bug disappears if I change the code to catch and throw instead of rethrow.

<details>
  <summary>Repro</summary>
  
  ```csharp
try
{
    await Test1();
}
catch (Exception ex)
{
    Console.WriteLine(ex.GetType().Name);
}

try
{
    await Test2();
}
catch (Exception ex)
{
    Console.WriteLine(ex.GetType().Name);
}

static async Task Test1()
{
    try
    {
        throw new Exception1();
    }
    catch (Exception1)
    {
        try
        {
            await Task.FromException(new Exception2());
        }
        catch (Exception2)
        {
            throw;
        }
    }
}

static async Task Test2()
{
    try
    {
        throw new Exception1();
    }
    catch (Exception1)
    {
        try
        {
            await Task.FromException(new Exception2());
        }
        catch (Exception2 ex)
        {
            throw ex;
        }
    }
}

public class Exception1 : Exception
{
}

public class Exception2 : Exception
{
}
  ```
  
</details>

It seems to also reproduce in sharplab when comparing "Default" vs "main":
Default: https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRj1UwAJUCBWAbjz2DAE88BvPUj8gTlIBUBTAM7ACACgCUtXAF88AYwCGwOQAtSogKIAPOfwAOwAJYB7KKX5bxrdpwpdRFgHQBxfsF6M9/CY4ByCgFt+STxZXHoma1xObj4hYHQJKTDFZTVNHX0jU3NLKJi7By0XNw8vH38gkJk6fAIANnIADnJGgWExK1w2aM4GZl6OHpiY4BUwYwB3Uih+ae1dAxMoTqkYsJjU1XUFrOWCLpjhkY5+mxPjk4KeVHrHADEJgN2l01FZ+czXqETxaqvSBsrlt0i9sj9DldLgDSGMJpM1lcgRwwmFiA1mq04sJfvk+pFBqRoaNxlMZnNSGD9klzoDaSCdl9wQdacT8QMAWyTqgbndHsZnkzlu8KVTTL9/kjaZslNsMotwehcpCLtKrnCyRZESdkXSauF8ABmchKsUrUggSlC0xRNFGk1WhXLJWWs22oA===

Main:
https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAAwCwAUEQIwX7EAE+VArANwUULICeFA3hXpDGATnoAVAKYBnBFQAUASnbkAvhQDGEBBoAW9eQFEAHhskAHBFAD2MepOOL+g4UxHyHAOgDikhOO5zSSVPADkIMEllCnVyTh5ncmFRCRkEQiUVWK0dfSNTCytbe0dE5LcPYx8/AKCQ8MjotQ5KKgA2RgAORg6pWQUncgEk4S5eEaFh5OSEXWRrAHd6GEklkzNLGxgBlWTY5Jy9A3XCrapB5KnpoTGXa6vr8rF8Ns8AMXmwE83beRW1go/GAZRRNR70faPQ55b5FYEXR4PcH0WbzBa7R6QoSxWK0dpdHqpWQgsqjBITehImZzRbLVb0WFnTJ3CEs6HHQFw84sqlk8bg3nXfDPV4faxfTlbP70xm2EFgzEsg7aI75DZwwglBH3JWPVG0hwY65Y1nNOKUADMjE1su29BADMltkSuMt1sd6q2modtpdQA

**Expected Behavior**:
Should output:
Exception2
Exception2

**Actual Behavior**:
Currently outputs:
Exception1
Exception2

**Comment by sharwell** [comment]
@jaredpar I marked this as a regression based on sharplab showing a change in behavior.

**Comment by AlekseyTs** [comment]
@jjonescz It looks like this is a regression from https://github.com/dotnet/roslyn/pull/70513. Would you like to take a look?

**Comment by EklipZgit** [comment]
this seems like something that maybe ought to make a servicing release no? We're over a month since the fix was merged to the compiler and stuff is still compiling throwing the wrong exceptions in live builds. While admittedly this is probably a pretty uncommon scenario for business logic to rely on, it still definitely has to be unexpectedly breaking a few applications in major ways in production, right?

**Comment by jjonescz** [comment]
It should be part of the next servicing release.

**Comment by kevinchalet** [comment]
For the record, @jaredpar mentioned on Twitter that this issue was fixed in VS 17.9.2: https://twitter.com/jaredpar/status/1762986759120667029.

The same fix is also expected to land in the .NET SDK (for CLI scenarios) in the near future: https://twitter.com/jaredpar/status/1762989260662280287.

**Comment by jaredpar** [comment]
Unfortunately the fix came through at a bad time in our servicing lifecycle so it's taking longer to get out than normal. The fix is still very much getting pushed through servicing but there are time constraints out of our control right now.

In the short term anyone impacted by this can add a reference to Microsoft.Net.Compilers.Toolset 4.9.2 to temporarily work around this. 

```csproj
    <PackageReference Include="Microsoft.Net.Compilers.Toolset" Version="4.9.2">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
```

**Note**: this package is **only** supported for temporary hot fixes like this. Once the fix is released via 8.0.2xx servicing please uninstall this package. It can and will break as new major / feature releases of the .NET SDK / MSBuild come out.

**Comment by jjonescz** [comment]
FYI: fix for this has been released in SDK 8.0.203 and VS 17.9.2
